### PR TITLE
Generalize RS0012 to handle more immutable collection types

### LINKED ISSUE #1284: Proposal: Generalize RS0012 for different immutable collection types
#### Analyzer package

Microsoft.NetCore.Analyzers

#### Analyzer

DoNotCallToImmutableArrayOnAnImmutableArrayValue

####  Proposal

`ToImmutableX` on an `ImmutableX` should also trigger a warning, not just when `X = Array`.

### PR DESCRIPTION
**Pull request by jamesqo**
Currently RS0012 only fires when you call `ToImmutableArray` on an `ImmutableArray`. This PR generalizes it so it also triggers when you call `ToImmutableHashSet` on an `ImmutableHashSet`, `ToImmutableList` on an `ImmutableList`, etc.

**List of changes**

- Generalize analyzer
- Update names of everything related to this analyzer (`DoNotCallToImmutableArrayOnAnImmutableArrayValue` -> `DoNotCallToImmutableCollectionOnAnImmutableCollectionValue`)
- Updated resource files (title & message) and XML documentation
- Updated tests to cover the new scenarios

Fixes #1284

### CODE REVIEW

File: src/Microsoft.NetCore.Analyzers/UnitTests/ImmutableCollections/DoNotCallToImmutableCollectionOnAnImmutableCollectionValueTests.cs [hunk 1]
```diff
@@ -0,0 +1,222 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (333fred) [inline src/Microsoft.NetCore.Analyzers/UnitTests/ImmutableCollections/DoNotCallToImmutableCollectionOnAnImmutableCollectionValueTests.cs]: Please add back the tests that use a custom extension method.

File: src/Microsoft.NetCore.Analyzers/Core/ImmutableCollections/DoNotCallToImmutableCollectionOnAnImmutableCollectionValue.cs [hunk 2]
```diff
@@ -39,30 +50,34 @@ public override void Initialize(AnalysisContext context)
 
             context.RegisterCompilationStartAction(compilationStartContext =>
             {
-                var immutableArrayType = compilationStartContext.Compilation.GetTypeByMetadataName(ImmutableArrayMetadataName);
-                if (immutableArrayType == null)
+                var compilation = compilationStartContext.Compilation;
+                if (compilation.GetTypeByMetadataName(ImmutableArrayMetadataName) == null)
```
Comment (333fred) [inline src/Microsoft.NetCore.Analyzers/Core/ImmutableCollections/DoNotCallToImmutableCollectionOnAnImmutableCollectionValue.cs]: That will work fine.

File: src/Roslyn.Diagnostics.Analyzers/Core/RoslynDiagnosticIds.cs [hunk 3]
```diff
@@ -15,7 +15,7 @@ internal static class RoslynDiagnosticIds
         // public const string OverrideObjectEqualsRuleId = "RS0009";                   // Now CA1815 => Microsoft.ApiDesignGuidelines.Analyzers.OverrideEqualsAndOperatorEqualsOnValueTypesAnalyzer
         // public const string DoNotUseVerbatimCrefsRuleId = "RS0010";                  // Now RS0010 => XmlDocumentationComments.Analyzers.AvoidUsingCrefTagsWithAPrefixAnalyzer
         // public const string CancellationTokenMustBeLastRuleId = "RS0011";            // Now CA1068 => Microsoft.ApiDesignGuidelines.Analyzers.CancellationTokenParametersMustComeLastAnalyzer
-        // public const string DoNotCallToImmutableArrayRuleId = "RS0012";              // Now RS0012 => System.Collections.Immutable.Analyzers.DoNotCallToImmutableArrayOnAnImmutableArrayValueAnalyzer
+        // public const string DoNotCallToImmutableArrayRuleId = "RS0012";              // Now RS0012 => System.Collections.Immutable.Analyzers.DoNotCallToImmutableCollectionOnAnImmutableCollectionValueAnalyzer
```
Comment (333fred) [inline src/Roslyn.Diagnostics.Analyzers/Core/RoslynDiagnosticIds.cs]: The diff you have here looks fine to me.

File: src/Unfactored/Roslyn/Core/RoslynDiagnosticsResources.resx [hunk 4]
```diff
@@ -189,8 +189,8 @@
   <data name="CancellationTokenMustBeLastMessage" xml:space="preserve">
     <value>Method '{0}' should take CancellationToken as the last parameter</value>
   </data>
-  <data name="DoNotCallToImmutableArrayMessage" xml:space="preserve">
-    <value>Do not call ToImmutableArray on an ImmutableArray&lt;T&gt; value.</value>
+  <data name="DoNotCallToImmutableCollectionMessage" xml:space="preserve">
```
Comment (333fred) [inline src/Unfactored/Roslyn/Core/RoslynDiagnosticsResources.resx]: This rule isn't used anywhere, but it should probably be updated anyway.

Comment (333fred) [inline src/Unfactored/Roslyn/Core/RoslynDiagnosticsResources.resx]: Actually, on second though, no need to worry about the unfactored folder or anything in it.

**Comment by 333fred** [comment]
@jamesqo looks like the unit tests are hanging.

**Comment by jamesqo** [comment]
@dotnet-bot retest prtest/windows/debug please

**Comment by jamesqo** [comment]
@333fred It's not the unit tests that are hanging, it appears to be this command:

```
D:\j\workspace\windows_debug---c729d7ab>.\CIBuild.cmd -configuration debug -clearCaches 
```

**Comment by jamesqo** [comment]
@dotnet-bot retest prtest/windows/debug please

**Comment by 333fred** [comment]
@jamesqo that's just this file: https://github.com/dotnet/roslyn-analyzers/blob/master/cibuild.cmd. It runs the build, you can run it locally.

**Comment by jamesqo** [comment]
@333fred I just ran the command locally, works fine for me. I think it could be something with your Debug CI machine.

**Comment by 333fred** [comment]
Ok, I'll investigate a bit later

**Comment by 333fred** [comment]
@jamesqo, the script appears to be hanging when downloading vswhere. Could you add `Write-Host "Downloading vswhere"` and `Write-Host "Downloaded vswhere"` around the Invoke-WebRequest here: https://github.com/dotnet/roslyn-analyzers/blob/master/build/build.ps1#L49 so we can be sure of that?

**Comment by jamesqo** [comment]
@333fred No problem. I just pushed a new commit with your requested changes to the build script.

**Comment by jamesqo** [comment]
@333fred Just curious about the `Unfactored` directory-- is any of the code there actually used? Why is it around?

**Comment by 333fred** [comment]
Unfactored was a bunch of prototype implementations, we'll delete it at some point. As for other analyzers, sure! Do note that the IOperation API is continuously changing at this point, and there are some known bugs that have been fixed in the next release of roslyn that will possibly affect you here. I don't have a full list, but this will be most of them (both open and closed: https://github.com/dotnet/roslyn/issues?utf8=%E2%9C%93&q=is%3Aissue%20label%3A%22New%20Feature%20-%20IOperation%22%20).

**Comment by jamesqo** [comment]
@333fred I opened a new issue for this: https://github.com/dotnet/roslyn-analyzers/issues/1289. Perhaps after a few days the problem has gone away, so I'll try running CI again. However, I remember seeing a green checkmark next to the release build a couple of days ago before I tried rebuilding, so if you believe me feel free to merge.

**Comment by jamesqo** [comment]
@dotnet-bot

retest prtest/windows/release please
retest prtest/windows/debug please

**Comment by jamesqo** [comment]
retest prtest/windows/debug please
retest prtest/windows/release please

**Comment by jamesqo** [comment]
@dotnet-bot test this please

**Comment by jamesqo** [comment]
@dotnet/roslyn-analysis @mavasani Can you please do a second review? Thanks.

File: src/Microsoft.NetCore.Analyzers/Core/ImmutableCollections/DoNotCallToImmutableCollectionOnAnImmutableCollectionValue.cs [hunk 5]
```diff
@@ -39,30 +48,36 @@ public override void Initialize(AnalysisContext context)
...
-                    if (invocation.TargetMethod?.Name != "ToImmutableArray")
+                    var targetMethod = invocation.TargetMethod;
+                    if (targetMethod == null || !ImmutableCollectionMetadataNames.TryGetValue(targetMethod.Name, out string metadataName))
                     {
                         return;
                     }
 
+                    Debug.Assert(!string.IsNullOrEmpty(metadataName));
+                    var immutableCollectionType = compilation.GetTypeByMetadataName(metadataName);
+
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/ImmutableCollections/DoNotCallToImmutableCollectionOnAnImmutableCollectionValue.cs]: Please either add a null check here for `immutableCollectionType`. The check above only verified if ImmutableArray is defined, but the user might be running against a custom system assembly that doesn't define other immutable types.

Comment (jamesqo) [reply src/Microsoft.NetCore.Analyzers/Core/ImmutableCollections/DoNotCallToImmutableCollectionOnAnImmutableCollectionValue.cs]: @mavasani I don't need to add tests for this, right? (The tests automatically reference the S.C.I assembly, so when I add that branch it'll never be covered.)

Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/ImmutableCollections/DoNotCallToImmutableCollectionOnAnImmutableCollectionValue.cs]: You can potentially do so by creating a compilation that doesn't reference System.Collections.Immutable.dll, and the test source defines a user defined type named `System.Collections.Immutable.ImmutableArray`

**Comment by mavasani** [comment]
@jamesqo Can you please fix the merge conflicts? I can merge once the tests are passing.

**Comment by jamesqo** [comment]
@mavasani I have fixed the merge conflicts.

**Comment by jinujoseph** [comment]
@jamesqo , can you address the build failure

**Comment by 333fred** [comment]
Thanks @jamesqo.

**Pull request merged**